osquery/events/pathset.h

```diff
@@ -11,149 +11,142 @@
     WriteLock lock(mset_lock_);
-    paths_.clear();
+    patterns_.clear();
   }
 
   bool empty() const {
     ReadLock lock(mset_lock_);
-    return paths_.empty();
+    return patterns_.empty();
   }
 
  private:
   typedef typename PathType::Path Path;
   typedef typename PathType::Compare Compare;
-  std::multiset<Path, Compare> paths_;
+  std::vector<Path> patterns_;
   mutable Mutex mset_lock_;
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view pattern,
+                        boost::string_view str) const {
+      if (pattern[0] == '*' && pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(1, pattern.size() - 2);
+        if (str.size() >= pattern.size() &&
+            str.find(pattern) != boost::string_view::npos) {
+          return true;
+        }
+        return false;
+      } else if (pattern[0] == '*') {
+        pattern = pattern.substr(1);
+        if (pattern.size() <= str.size()) {
+          str = str.substr(str.size() - pattern.size());
+        } else {
+          return false;
+        }
+      } else if (pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(0, pattern.size() - 1);
+        if (pattern.size() <= str.size()) {
+          str = str.substr(0, pattern.size());
+        } else {
+          return false;
+        }
+      }
+
+      return (pattern == str);
+    }
+
     bool operator()(const Path& lhs, const Path& rhs) const {
-      size_t psize = (lhs.size() < rhs.size()) ? lhs.size() : rhs.size();
+      auto psize = std::min(lhs.size(), rhs.size());
       unsigned ndx;
       for (ndx = 0; ndx < psize; ++ndx) {
-        if (lhs[ndx] == "**" || rhs[ndx] == "**") {
-          return false;
+        if (lhs[ndx] == "**") {
+          return true;
         }
 
-        if (lhs[ndx] == "*" || rhs[ndx] == "*") {
+        if (lhs[ndx] == "*") {
           continue;
         }
 
-        int rc = lhs[ndx].compare(rhs[ndx]);
-
-        if (rc > 0) {
+        // compare with partial patterns
+        if (compareStrings(lhs[ndx], rhs[ndx]) == true) {
+          continue;
+        } else {
           return false;
         }
-
-        if (rc < 0) {
-          return true;
-        }
-      }
-
-      if ((ndx == rhs.size() && rhs[ndx - 1] == "*") ||
-          (ndx == lhs.size() && lhs[ndx - 1] == "*")) {
-        return false;
       }
 
-      return (lhs.size() < rhs.size());
+      return (lhs.size() == rhs.size());
     }
   };
 
-  static Path createPath(const std::string& str) {
+  static Path createPath(std::string str) {
     boost::char_separator<char> sep{"/"};
     tokenizer tokens(str, sep);
     Path path;
 
     if (str == "/") {
-      path.push_back("");
+      path.push_back("/");
     }
 
     for (std::string component : tokens) {
```

Comment:
I see these two statements are same. If Library is a folder then statement should be same. OS will figure out  if Library is folder or not, even if you provide '/' or not. Form exclude perspective 
'/This/Path' or '/This/Path/' and '/This/Path/is' are two different objects. If I say filter out  '/This/Path/' this should not mean filter out '/This/Path/is' as well since for this we have '/This/Path/%'.  
Again there should not be any confusion about path ending with '/' . Since OS will decide if provided object is directory or not, we can not give this a new meaning at application layer.